### PR TITLE
=act Reset receiveTimeoutData to emptyReceiveTimeoutData when cancel

### PR DESCRIPTION
**Pull request by He-Pin**
References https://github.com/akka/akka/issues/31055

### CODE REVIEW

File: akka-docs/src/test/java/jdocs/actor/ActorDocTest.java [hunk 1]
```diff
@@ -581,6 +581,7 @@ public void using_receiveTimeout() {
         myActor.tell("Hello", getRef());
         expectMsgEquals("Hello world");
         expectMsgEquals("timeout");
+        expectNoMessage(Duration.ofSeconds(2));
```
Comment (hepin1989) [inline akka-docs/src/test/java/jdocs/actor/ActorDocTest.java:584]: add this method to test will not receive Timeout after cancel the receive timeout.

Comment (johanandren) [reply akka-docs/src/test/java/jdocs/actor/ActorDocTest.java:584]: `expectNoMessage` will always wait for the full duration, a permanent 2 seconds is a lot to add to a test, would be better create a separate test with a shorter timeout for this, but otoh it is perhaps too easy to cause a race condition in the test with a short timeout (cancelling only prevents future receive timeout, not one that already triggered but was not yet processed by the actor)

File: akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala [hunk 2]
```diff
@@ -60,12 +60,12 @@ private[akka] trait ReceiveTimeout { this: ActorCell =>
...
   protected def cancelReceiveTimeoutIfNeeded(message: Any): (Duration, Cancellable) = {
     if (hasTimeoutData && !message.isInstanceOf[NotInfluenceReceiveTimeout])
-      cancelReceiveTimeout()
+      cancelReceiveTimeoutTask()
 
     receiveTimeoutData
   }
 
-  override final def cancelReceiveTimeout(): Unit =
+  private def cancelReceiveTimeoutTask(): Unit =
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala:68]: Wouldn't it be more clear we rather remove the method implementation in `akka.actor.AbstractActor.ActorContext#cancelReceiveTimeout` and make that abstract (as that currently has a different logic now, leaving the cancellable around)?

Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala:68]: wow, so this is kind of accidentally overriding the method in `AbstractActor`

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-docs/src/test/java/jdocs/actor/ActorDocTest.java [hunk 3]
```diff
@@ -585,6 +585,37 @@ public void using_receiveTimeout() {
...
+  }
+
+  @Test
+  public void cancel_receiveTimeout() {
+    final ActorRef myActor = system.actorOf(Props.create(ReceiveTimeoutWithCancelActor.class));
+    new TestKit(system) {
+      {
+        myActor.tell("Hello", getRef());
+        expectMsgEquals("Hello world");
+        expectNoMessage(Duration.ofMillis(400));
```
[the remark this answers is not in the document]
Comment (hepin1989) [reply akka-docs/src/test/java/jdocs/actor/ActorDocTest.java:614]: @johanandren I have updated the PR.

File: akka-docs/src/test/java/jdocs/actor/ActorDocTest.java [hunk 4]
```diff
@@ -585,6 +585,37 @@ public void using_receiveTimeout() {
...
+    }
+
+    @Override
+    public Receive createReceive() {
+      return receiveBuilder()
+          .matchEquals(
+              "Hello",
+              s -> {
+                // To set in a response to a message
+                getContext().setReceiveTimeout(Duration.ofMillis(200));
```
Comment (johanandren) [inline akka-docs/src/test/java/jdocs/actor/ActorDocTest.java:600]: Shouldn't it (also) cancel the timeout, since that is what we are testing?

Comment (hepin1989) [reply akka-docs/src/test/java/jdocs/actor/ActorDocTest.java:600]: oops, I was buying stocks when coding this... will update then

**Review by johanandren — APPROVED** [review]
LGTM, thanks @hepin1989 !

**Pull request merged**